docs/decisions/0021-transaction-aggregates.rst

```diff
@@ -0,0 +1,135 @@
+from a a deployment standpoint, and it helps isolate/secure the transaction-level data in the
+enterprise-subsidy service. However, this design has several tradeoffs:
+
+1. Tight "runtime" dependencies exist - the ``can_redeem`` business logic and the redemption
+   workflow cannot execute at all without both the enterprise-access and enterprise-subsidy services
+   running. Furthermore, that data in each service is highly coupled - policy records in enterprise-access
+   *must*  relate to a well-formed and active *subsidy* record in the enterprise-subsidy service in
+   order for the redemption workflow to execute.
+2. These runtime dependencies utilize synchronous HTTP requests for both the ``can_redeem`` and
+   ``redeem`` workflows. This makes the system as a whole less scalable, less elastic,
+   and less performant, because each of these workflows mush synchronously block on requests from
+   enterprise-access to enterprise-subsidy.
+   
+
+Decision
+========
+Build and maintain *aggregate* redemption models in the enterprise-access service. The aggregated
+metrics we're most interested in right now are total spend (the sum of transaction quantities) and the count
+of transactions.
+The aggregated *dimensions* we're most interested in right now are as follows:
+
+1. count/spend metrics at the **subsidy** level - based on ``subsidy_uuid``
+2. count/spend metrics at the **policy** level - based on ``subsidy_access_policy_uuid``
+3. count/spend metrics at the **(policy, learner)** level - based on (policy id, learner id)
+
+Additionally, we may eventually want aggregates at the **(policy, learner, group)** level, to support
+business record configurations where multiple groups relate to a single policy.
+
+Once these aggregates are in place, replace requests to the enterpirse-subsidy service
+in the ``can_redeem`` and ``redeem`` flows with database queries against these aggregates.
+
+Maintaining these aggregates depends on sourcing transaction and reversal data as follows:
+
+1. Starting from a zero-state in which Aggregate models exist but are not yet populated (backpopulation).
+2. When transactions are committed, any related aggregate records should be updated to reflect it.
+3. When reversals are committed, any related aggregate records should be updated to reflect it.
+
+Inter-service Communication
+---------------------------
+There are several options for how transaction/reversal data can make its way to the aggregates
+layer in the enterprise-access service:
+
+1. Via management commands that reads transaction and reversal records via the enterprise-subsidy
+   REST API.  This approach may work for backpopulation, but is generally disfavorable because
+   it won't help us keep the aggregate records up-to-date in a timely manner (we'd have to run
+   it on some cron/schedule).
+2. Updating the aggregates based on the ``redeem`` response body (on successful redemption).
```

Comment:
nit: If we will not use option 2 or 3 should they be moved to the "Alternatives Considered" block?

Reply:
I view "Alternatives considered" as a place to document different approaches to the general problem.  This is more like the rejected alternatives to a sub-problem: having accepted that we'll maintain aggregates in enterprise-access, how will the data get there?

Reply:
No strong feelings.